mapstory/apps/thumbnails/test_storylayer_thumbnail.py

```diff
@@ -0,0 +1,103 @@
+from mapstory.tests.integration.geogig_uploader import GeoGigUploaderBase
+from django.test import TestCase
+import os
+
+from geonode.layers.models import Layer
+from mapstory.apps.thumbnails.tasks import CreateStoryLayerThumbnailTask
+from PIL import Image
+from StringIO import StringIO
+
+
+# python manage.py test --noinput --nocapture  mapstory.apps.thumbnails.test_storylayer_thumbnail
+class TestStoryLayerThumbnailTask(GeoGigUploaderBase, TestCase):
+    # note - while this is running, there is likely a thumbnail generation task occuring in the background
+    # (kicked off by the importer)
+    #   So, the layer may or may not have a thumbnail attached to it (and might have one in the future).
+    def test_withFeatures(self):
+        layer = self.fully_import_file(os.path.realpath('mapstory/tests/sampledata/'), 'railroads.zip', 'YEAR')
+        thumb_generator = CreateStoryLayerThumbnailTask()
+
+        # this layer has features in it
+        self.assertTrue(thumb_generator.has_features(layer))
+
+        # validate bounding box and time
+        bounding_box, timepositions = thumb_generator.retreive_WMS_metadata(layer)
+        self.assertTrue(abs(bounding_box[0] - -123.035331726) < 0.0000001)
+        self.assertTrue(abs(bounding_box[1] - 28.609628677368164) < 0.0000001)
+        self.assertTrue(abs(bounding_box[2] - -67.27764129638672) < 0.0000001)
+        self.assertTrue(abs(bounding_box[3] - 46.011295318603516) < 0.0000001)
+
+        self.assertEqual(len(timepositions), 6)
+        self.assertEqual(timepositions[0], '1840-01-01T00:00:00.000Z')
+        self.assertEqual(timepositions[5], '1870-01-01T00:00:00.000Z')
+
+        # this shouldn't throw an error
+        thumb_generator.run(layer.pk, overwrite=True)
+        layer = Layer.objects.get(pk=layer.pk)  # reload from disk (the task will change it in a different context)
+
+        # this should not be the default
+        self.assertTrue(layer.has_thumbnail())
+
+        # imbedded (in Layer#Links) should be the same as saved in Layer#thumbnail_url)
+        self.assertEqual(layer.thumbnail_url, layer.get_thumbnail_url())
+
+        # url is correct
+        self.assertTrue(layer.thumbnail_url.endswith(thumb_generator.get_official_thumbnail_name(layer)))
+
+        # verify that setting thumbnail to default works
+        thumb_generator.set_layer_thumbnail_default(layer)
+        self.assertFalse(layer.has_thumbnail())
+        self.assertEqual(layer.thumbnail_url, layer.get_thumbnail_url())
+
+        # verify image - quick check to see if the size is correct (which means it generated an image)
+        #  TO DO: add test to check actual image (note -- highly unstable because it relies on Geoserver style
+        #         and basemap to NOT change).
```

Comment:
Any ideas how to do this in a less fragile manner?

Reply:
@Coop56 ?

Reply:
Is there an easy way to regenerate the image if we change the styles?

Reply:
I can put in a more explicit image test, but its likely to be fragile.  However, maintaining shouldn't be too hard - if the image changes, then just regenerate the test image and commit the changed  test image.

In a more production way (i.e. if the default style changes), you'd just have to do something like this;

foreach layer.pk in layers
     task = CreateStoryLayerThumbnailTask()
     task.delay(pk, overwrite=True) 

And it will schedule a thumbnail generation for each layer.